README.md
[![Build Status](https://travis-ci.org/jscs-dev/jscs-spellcheck.svg?branch=master)](https://travis-ci.org/jscs-dev/jscs-spellcheck)
[![Windows CI](https://ci.appveyor.com/api/projects/status/github/jscs-dev/jscs-spellcheck?svg=true)](https://ci.appveyor.com/project/jscs-dev/jscs-spellcheck/branch/master)
[![Coverage Status](https://img.shields.io/coveralls/jscs-dev/jscs-spellcheck.svg?style=flat)](https://coveralls.io/r/jscs-dev/jscs-spellcheck?branch=master)
[![Dependency Status](https://david-dm.org/jscs-dev/jscs-spellcheck.svg?theme=shields.io&style=flat)](https://david-dm.org/jscs-dev/jscs-spellcheck)
[![devDependency Status](https://david-dm.org/jscs-dev/jscs-spellcheck/dev-status.svg?theme=shields.io&style=flat)](https://david-dm.org/jscs-dev/jscs-spellcheck#info=devDependencies)

**JSCS Spellcheck** — Spellcheck plugin for [JSCS](https://github.com/jscs-dev/node-jscs/).

The majority of time spent programming is not spent typing, but in reading and
maintaining existing code. It is important that other programmers be able to
understand exactly what is going on in a complex system.

Abbreviations and contractions can make code harder to understand. They can be
ambiguous and arbitrary. Abbreviations cannot be distinguished from
misspellings. Today, there are free integrated development environments which
provide word auto-completion and line wrapping; there is no practical advantage
to using abbreviations in code.

This JSCS plugin checks for words that can't be found in a dictionary, and tells
you where they are so that you can spell them correctly. You can choose which
dictionaries and languages to use. You can add more words and ignore existing
ones. You can define exceptions for names used by 3rd parties. You can even
restrict a word's usage to identifiers or property names only.

## Installation

`jscs-spellcheck` can be installed using NPM and requires
[jscs](https://github.com/jscs-dev/node-jscs/#installation).

Install it globally if you are using globally installed `jscs`:

    npm install jscs-spellcheck --global

But better install it into your project:

    npm install jscs-spellcheck --save-dev

## Usage

To use, add these lines to your `.jscsrc` configuration file:

```json
{
    "plugins": [ "jscs-spellcheck" ],
    "requireDictionaryWords": {
        "dictionaries": [ "english", "american" ]
    }
}
```

## Rules

### requireDictionaryWords

Only allow words defined in a dictionary or by you.

English language support is installed by default. To add additional
languages, go to the installation directory of JSCS and run `npm install
wordlist-LANGUAGE`, where LANGUAGE is the name of your language.

Type: `Boolean` or `Object`

Values:
 - `true`: use the `"english"` dictionary
 - `Object`:
   - `dictionaries`: (default `["english"]`) array of dictionary names including
     `"english"`, `"american"`, `"british"` and `"canadian"`
   - `allowWords`: additional words allowed anywhere
   - `allowWordsInIdentifiers`: additional words allowed only in identifiers
   - `allowWordsInProperties`: additional words allowed only in properties
   - `allowNames`: whole names ignored by spellcheck
   - `allowNamesAsIdentifiers`: whole names ignored by spellcheck when used as identifiers
   - `allowNamesAsProperties`: whole names ignored by spellcheck when used as properties
   - `excludeWords`: words to exclude from the dictionaries

#### Example

```js
"requireDictionaryWords": true

"requireDictionaryWords": {
    "dictionaries": [ "english", "american" ],
    "allowWords": [ "transclude" ],
    "allowWordsInProperties": [ "chmod" ],
    "allowNamesAsIdentifiers": [ "$stateParams", "util" ],
    "allowNamesAsProperties": [ "src" ],
    "excludeWords": [ "i" ]
}
```

##### Valid for mode `true`

```js
var number = 1;
object['source'] = 2;
object.source = 3;
fileDirectory = 4;
```

##### Invalid for mode `true`

```js
var num = 1;
obj['src'] = 2;
obj.src = 3;
fileDir = 4;
```

##### Valid for mode `"dictionaries": [ "american" ]`, invalid for `"british"`

```js
var color = 'papayawhip';
```

##### Valid for mode `"dictionaries": [ "british" ]`, invalid for `"american"`

```js
var colour = 'papayawhip';
```

##### Valid for mode `"allowWords": [ "transclude" ]`

```js
var transclude = function() {};
var transcludeFunction = function() {};
return { transclude: function() {} };
```

##### Valid for mode `"allowWordsInProperties": [ "chmod" ]`

```js
var mode = 0777;
fs.chmod('/', mode, function(error) {});
fs.chmodSync('/', mode);
```

##### Invalid for mode `"allowWordsInProperties": [ "chmod" ]`

```js
var chmod = 0777;
```

##### Valid for mode `"allowNamesAsIdentifiers": [ "$stateParams", "util" ]`

```js
var util = require('util');
function Controller($stateParams) {}
```

##### Invalid for mode `"allowNamesAsIdentifiers": [ "$stateParams", "util" ]`

```js
var stringUtil = {};
var params = {};
```

##### Valid for mode `"allowNamesAsProperties": [ "src" ]`

```js
element.src = 'https://youtu.be/dQw4w9WgXcQ';
```

##### Invalid for mode `"allowNamesAsProperties": [ "src" ]`

```js
var data = { videoSrc: 'youtube' };
```

##### Invalid for mode `"excludeWords": [ "i" ]`

```js
for (var i = 0; i < array.length; i++) {}
```

#### Error Suppression

You can also allow (and later disallow) words on a per-file (or per-line)
basis:

```js
// jscs:allowWords concat, dest, dist, src
grunt.initConfig({
    concat: {
        dist: {
            src: ['src/*.js'],
            dest: 'dist/scripts.js'
        }
    }
});
```

```js
// jscs:allowNamesAsIdentifiers EOL
var EOL = require('os').EOL;
// jscs:disallowNamesAsIdentifiers EOL

if (age > 80) var reachingEol = true; // invalid
```


CHANGELOG.md
## Version 1.0.0

### Overview
Initial release.


test/specs/rules/require-dictionary-words.js
var Checker = require('jscs/lib/checker');
var assert = require('assert');

describe('rules/require-dictionary-words', function() {
    var checker;

    beforeEach(function() {
        checker = new Checker();
        checker.registerDefaultRules();
        checker.configure({ plugins: ['./lib/index.js'] });
    });

    describe('true', function() {
        beforeEach(function() {
            checker.configure({ requireDictionaryWords: true });
        });

        it('should report non-words', function() {
            assert(checker.checkString('asdf = 1;').getErrorCount() === 1);
            assert(checker.checkString('var asdf = 1;').getErrorCount() === 1);
            assert(checker.checkString('let asdf = 1;').getErrorCount() === 1);
            assert(checker.checkString('const asdf = 1;').getErrorCount() === 1);
            assert(checker.checkString('function asdf(jkl) {var asdf = 1;}').getErrorCount() === 3);
            assert(checker.checkString('var asdf = function asdf() {};').getErrorCount() === 2);
            assert(checker.checkString('for(asdf = 0; asdf<1; asdf++){asdf;}').getErrorCount() === 1);
            assert(checker.checkString('object["jkl"] = 1;').getErrorCount() === 1);
            assert(checker.checkString('object.jkl = 1;').getErrorCount() === 1);
            assert(checker.checkString('object = {jkl: 1};').getErrorCount() === 1);
            assert(checker.checkString('object = {"jkl": 1};').getErrorCount() === 1);
            assert(checker.checkString('asdf: void 0;').getErrorCount() === 1);

            assert(checker.checkString('JKL = 1;').getErrorCount() === 1);
        });

        it('should report multiple non-words', function() {
            assert(checker.checkString('asdfAsdf = 1;').getErrorCount() === 2);
            assert(checker.checkString('asdfASDF = 1;').getErrorCount() === 2);
            assert(checker.checkString('asdfASDFAsdf = 1;').getErrorCount() === 3);
            assert(checker.checkString('asdf_asdf = 1;').getErrorCount() === 2);
        });

        it('should ignore numbers and symbols', function() {
            assert(checker.checkString('$asdfAsdf = 1;').getErrorCount() === 2);
            assert(checker.checkString('asdf$Asdf = 1;').getErrorCount() === 2);
            assert(checker.checkString('asdfAsdf$ = 1;').getErrorCount() === 2);
            assert(checker.checkString('asdf9Asdf = 1;').getErrorCount() === 2);
            assert(checker.checkString('asdfAsdf9 = 1;').getErrorCount() === 2);
        });

        it('should not report non-"instantiation" usages', function() {
            assert(checker.checkString('asdf').isEmpty());
            assert(checker.checkString('asdf.jkl').isEmpty());
            assert(checker.checkString('asdf(jkl)').isEmpty());
            assert(checker.checkString('asdf.jkl(jkl)').isEmpty());
            assert(checker.checkString('value = asdf').isEmpty());
            assert(checker.checkString('var value = asdf').isEmpty());
            assert(checker.checkString('object.property = asdf').isEmpty());
            assert(checker.checkString('if(asdf){}').isEmpty());
            assert(checker.checkString('object[jkl] = 1;').isEmpty());
        });

        it('should not report defined words', function() {
            assert(checker.checkString('good = 1;').isEmpty());
        });

        it('should not report multiple defined words', function() {
            assert(checker.checkString('goodGood = 1;').isEmpty());
        });
    });

    describe('esnext', function() {
        beforeEach(function() {
            checker.configure({ requireDictionaryWords: true, esnext: true });
        });

        it('should report non-words in es6', function() {
            assert(checker.checkString('object = {asdf, jkl() {}, [0 + 1]: 2};').getErrorCount() === 2);
            assert(checker.checkString('class Asdf { jkl() {} }').getErrorCount() === 2);
        });

        it('should not report defined words in es6', function() {
            assert(checker.checkString('object = {good, good() {}, [0 + 1]: 2};').isEmpty());
            assert(checker.checkString('class Good { good() {} }').isEmpty());
        });
    });

    describe('allowWords', function() {
        beforeEach(function() {
            checker.configure({
                requireDictionaryWords: {
                    allowWords: ['asdf', 'jkl']
                }
            });
        });

        it('should not report allowed words', function() {
            assert(checker.checkString('asdf = 1;').isEmpty());
            assert(checker.checkString('object.jkl = 1;').isEmpty());
            assert(checker.checkString('asdfAsdf = 1;').isEmpty());
            assert(checker.checkString('object.jklJkl = 1;').isEmpty());
        });
    });

    describe('allowNames', function() {
        beforeEach(function() {
            checker.configure({
                requireDictionaryWords: {
                    allowNames: ['asdf', 'jkl']
                }
            });
        });

        it('should report names used as parts of names', function() {
            assert(checker.checkString('asdfAsdf = 1;').getErrorCount() === 2);
            assert(checker.checkString('object.jklJkl = 1;').getErrorCount() === 2);
        });

        it('should not report allowed names', function() {
            assert(checker.checkString('asdf = 1;').isEmpty());
            assert(checker.checkString('object.jkl = 1;').isEmpty());
        });
    });

    describe('allowWordsInIdentifiers', function() {
        beforeEach(function() {
            checker.configure({
                requireDictionaryWords: {
                    allowWordsInIdentifiers: ['asdf', 'jkl']
                }
            });
        });

        it('should report non-words in properties', function() {
            assert(checker.checkString('object.jkl = 1;').getErrorCount() === 1);
            assert(checker.checkString('object.jklJkl = 1;').getErrorCount() === 2);
        });

        it('should not report allowed words in identifiers', function() {
            assert(checker.checkString('asdf = 1;').isEmpty());
            assert(checker.checkString('asdfAsdf = 1;').isEmpty());
        });
    });

    describe('allowWordsInProperties', function() {
        beforeEach(function() {
            checker.configure({
                requireDictionaryWords: {
                    allowWordsInProperties: ['asdf', 'jkl']
                }
            });
        });

        it('should report non-words in identifiers', function() {
            assert(checker.checkString('asdf = 1;').getErrorCount() === 1);
            assert(checker.checkString('asdfAsdf = 1;').getErrorCount() === 2);
        });

        it('should not report allowed words in properties', function() {
            assert(checker.checkString('object.jkl = 1;').isEmpty());
            assert(checker.checkString('object.jklJkl = 1;').isEmpty());
        });
    });

    describe('allowNamesAsIdentifiers', function() {
        beforeEach(function() {
            checker.configure({
                requireDictionaryWords: {
                    allowNamesAsIdentifiers: ['asdf', 'jkl']
                }
            });
        });

        it('should report non-names and properties', function() {
            assert(checker.checkString('asdfAsdf = 1;').getErrorCount() === 2);
            assert(checker.checkString('object.jkl = 1;').getErrorCount() === 1);
            assert(checker.checkString('object.jklJkl = 1;').getErrorCount() === 2);
        });

        it('should not report allowed names as identifiers', function() {
            assert(checker.checkString('asdf = 1;').isEmpty());
        });
    });

    describe('allowNamesAsProperties', function() {
        beforeEach(function() {
            checker.configure({
                requireDictionaryWords: {
                    allowNamesAsProperties: ['asdf', 'jkl']
                }
            });
        });

        it('should report non-names and identifiers', function() {
            assert(checker.checkString('asdf = 1;').getErrorCount() === 1);
            assert(checker.checkString('asdfAsdf = 1;').getErrorCount() === 2);
            assert(checker.checkString('object.jklJkl = 1;').getErrorCount() === 2);
        });

        it('should not report allowed names as properties', function() {
            assert(checker.checkString('object.jkl = 1;').isEmpty());
        });
    });

    describe('excludeWords', function() {
        beforeEach(function() {
            checker.configure({
                requireDictionaryWords: {
                    excludeWords: ['good']
                }
            });
        });

        it('should report excluded words', function() {
            assert(checker.checkString('good = 1;').getErrorCount() === 1);
            assert(checker.checkString('object.good = 1;').getErrorCount() === 1);
        });
    });

    describe('dictionaries', function() {
        describe('english', function() {
            beforeEach(function() {
                checker.configure({
                    requireDictionaryWords: {
                        dictionaries: ['english']
                    }
                });
            });

            it('should report non-words', function() {
                assert(checker.checkString('color = 1;').getErrorCount() === 1);
            });
        });

        describe('english, american', function() {
            beforeEach(function() {
                checker.configure({
                    requireDictionaryWords: {
                        dictionaries: ['english', 'american']
                    }
                });
            });

            it('should not report defined words', function() {
                assert(checker.checkString('color = 1;').isEmpty());
            });
        });

        describe('missing dictionary', function() {
            it('should throw an informative error', function() {
                var missingLanguage = 'elvish';
                assert.throws(function() {
                    checker.configure({
                        requireDictionaryWords: {
                            dictionaries: [missingLanguage]
                        }
                    });
                }, new RegExp('wordlist-' + missingLanguage));
            });
        });
    });

    describe('inline configuration', function() {
        beforeEach(function() {
            checker.configure({ requireDictionaryWords: true });
        });

        it('should split', function() {
            assert(checker.checkString([
                '// jscs:allowWords',
                'asdf = 1',
                'jkl = 1'
            ].join('\n')).getErrorCount() === 2);
            assert(checker.checkString([
                '// jscs:allowWords asdf',
                'asdf = 1',
                'jkl = 1'
            ].join('\n')).getErrorCount() === 1);
            assert(checker.checkString([
                '// jscs:allowWords asdf, jkl',
                'asdf = 1',
                'jkl = 1'
            ].join('\n')).isEmpty());
            assert(checker.checkString([
                '// jscs:allowWords asdf, jkl,',
                'asdf = 1',
                'jkl = 1'
            ].join('\n')).isEmpty());
            assert(checker.checkString([
                '/* jscs:allowWords',
                ' asdf, jkl */',
                'asdf = 1',
                'jkl = 1'
            ].join('\n')).isEmpty());
            assert(checker.checkString([
                '/* jscs:allowWords',
                ' asdf,',
                ' jkl */',
                'asdf = 1',
                'jkl = 1'
            ].join('\n')).isEmpty());
        });

        it('should later allow and later disallow inline', function() {
            assert(checker.checkString([
                'asdf = 1',
                '// jscs:allowWords asdf, jkl',
                'asdf = 1',
                '// jscs:disallowWords asdf, jkl',
                'asdf = 1'
            ].join('\n')).getErrorCount() === 2);

            assert(checker.checkString([
                'asdf = 1',
                'jkl = 1',
                '// jscs:allowWords asdf, jkl',
                'asdf = 1',
                'jkl = 1',
                '// jscs:disallowWords asdf, jkl',
                'asdf = 1',
                'jkl = 1'
            ].join('\n')).getErrorCount() === 4);
        });

        it('should allow words inline', function() {
            assert(checker.checkString([
                '// jscs:allowWords asdf, jkl',
                'asdf = 1',
                'asdfAsdf = 1',
                'object.jkl = 1',
                'object.jklJkl = 1'
            ].join('\n')).isEmpty());
        });

        it('should allow words in identifiers inline', function() {
            assert(checker.checkString([
                '// jscs:allowWordsInIdentifiers asdf, jkl',
                'asdf = 1',
                'asdfAsdf = 1'
            ].join('\n')).isEmpty());
            assert(checker.checkString([
                '// jscs:allowWordsInIdentifiers asdf, jkl',
                'object.jkl = 1',
                'object.jklJkl = 1'
            ].join('\n')).getErrorCount() === 3);
        });

        it('should allow words in properties inline', function() {
            assert(checker.checkString([
                '// jscs:allowWordsInProperties asdf, jkl',
                'object.jkl = 1',
                'object.jklJkl = 1'
            ].join('\n')).isEmpty());
            assert(checker.checkString([
                '// jscs:allowWordsInProperties asdf, jkl',
                'asdf = 1',
                'asdfAsdf = 1'
            ].join('\n')).getErrorCount() === 3);
        });

        it('should allow names inline', function() {
            assert(checker.checkString([
                '// jscs:allowNames asdf, jkl',
                'asdf = 1',
                'object.jkl = 1'
            ].join('\n')).isEmpty());
            assert(checker.checkString([
                '// jscs:allowNames asdf, jkl',
                'asdfAsdf = 1',
                'object.jklJkl = 1'
            ].join('\n')).getErrorCount() === 4);
        });

        it('should allow names as identifiers inline', function() {
            assert(checker.checkString([
                '// jscs:allowNamesAsIdentifiers asdf, jkl',
                'asdf = 1'
            ].join('\n')).isEmpty());
            assert(checker.checkString([
                '// jscs:allowNamesAsIdentifiers asdf, jkl',
                'asdfAsdf = 1',
                'object.jkl = 1',
                'object.jklJkl = 1'
            ].join('\n')).getErrorCount() === 5);
        });

        it('should allow names as properties inline', function() {
            assert(checker.checkString([
                '// jscs:allowNamesAsProperties asdf, jkl',
                'object.jkl = 1'
            ].join('\n')).isEmpty());
            assert(checker.checkString([
                '// jscs:allowNamesAsProperties asdf, jkl',
                'asdf = 1',
                'asdfAsdf = 1',
                'object.jklJkl = 1'
            ].join('\n')).getErrorCount() === 5);
        });
    });
});


lib/rules/require-dictionary-words.js
/**
 * Only allow words defined in a dictionary or by you.
 *
 * English language support is installed by default. To add additional
 * languages, go to the installation directory of JSCS and run `npm install
 * wordlist-LANGUAGE`, where LANGUAGE is the name of your language.
 *
 * Type: `Boolean` or `Object`
 *
 * Values:
 *  - `true`: use the `"english"` dictionary
 *  - `Object`:
 *    - `dictionaries`: (default `["english"]`) array of dictionary names including
 *      `"english"`, `"american"`, `"british"` and `"canadian"`
 *    - `allowWords`: additional words allowed anywhere
 *    - `allowWordsInIdentifiers`: additional words allowed only in identifiers
 *    - `allowWordsInProperties`: additional words allowed only in properties
 *    - `allowNames`: whole names ignored by spellcheck
 *    - `allowNamesAsIdentifiers`: whole names ignored by spellcheck when used as identifiers
 *    - `allowNamesAsProperties`: whole names ignored by spellcheck when used as properties
 *    - `excludeWords`: words to exclude from the dictionaries
 *
 * #### Example
 *
 * ```js
 * "requireDictionaryWords": true
 *
 * "requireDictionaryWords": {
 *     "dictionaries": [ "english", "american" ],
 *     "allowWords": [ "transclude" ],
 *     "allowWordsInProperties": [ "chmod" ],
 *     "allowNamesAsIdentifiers": [ "$stateParams", "util" ],
 *     "allowNamesAsProperties": [ "src" ],
 *     "excludeWords": [ "i" ]
 * }
 * ```
 *
 * ##### Valid for mode `true`
 *
 * ```js
 * var number = 1;
 * object['source'] = 2;
 * object.source = 3;
 * fileDirectory = 4;
 * ```
 *
 * ##### Invalid for mode `true`
 *
 * ```js
 * var num = 1;
 * obj['src'] = 2;
 * obj.src = 3;
 * fileDir = 4;
 * ```
 *
 * ##### Valid for mode `"dictionaries": [ "american" ]`, invalid for `"british"`
 *
 * ```js
 * var color = 'papayawhip';
 * ```
 *
 * ##### Valid for mode `"dictionaries": [ "british" ]`, invalid for `"american"`
 *
 * ```js
 * var colour = 'papayawhip';
 * ```
 *
 * ##### Valid for mode `"allowWords": [ "transclude" ]`
 *
 * ```js
 * var transclude = function() {};
 * var transcludeFunction = function() {};
 * return { transclude: function() {} };
 * ```
 *
 * ##### Valid for mode `"allowWordsInProperties": [ "chmod" ]`
 *
 * ```js
 * var mode = 0777;
 * fs.chmod('/', mode, function(error) {});
 * fs.chmodSync('/', mode);
 * ```
 *
 * ##### Invalid for mode `"allowWordsInProperties": [ "chmod" ]`
 *
 * ```js
 * var chmod = 0777;
 * ```
 *
 * ##### Valid for mode `"allowNamesAsIdentifiers": [ "$stateParams", "util" ]`
 *
 * ```js
 * var util = require('util');
 * function Controller($stateParams) {}
 * ```
 *
 * ##### Invalid for mode `"allowNamesAsIdentifiers": [ "$stateParams", "util" ]`
 *
 * ```js
 * var stringUtil = {};
 * var params = {};
 * ```
 *
 * ##### Valid for mode `"allowNamesAsProperties": [ "src" ]`
 *
 * ```js
 * element.src = 'https://youtu.be/dQw4w9WgXcQ';
 * ```
 *
 * ##### Invalid for mode `"allowNamesAsProperties": [ "src" ]`
 *
 * ```js
 * var data = { videoSrc: 'youtube' };
 * ```
 *
 * ##### Invalid for mode `"excludeWords": [ "i" ]`
 *
 * ```js
 * for (var i = 0; i < array.length; i++) {}
 * ```
 *
 * #### Error Suppression
 *
 * You can also allow (and later disallow) words on a per-file (or per-line)
 * basis:
 *
 * ```js
 * // jscs:allowWords concat, dest, dist, src
 * grunt.initConfig({
 *     concat: {
 *         dist: {
 *             src: ['src/*.js'],
 *             dest: 'dist/scripts.js'
 *         }
 *     }
 * });
 * ```
 *
 * ```js
 * // jscs:allowNamesAsIdentifiers EOL
 * var EOL = require('os').EOL;
 * // jscs:disallowNamesAsIdentifiers EOL
 *
 * if (age > 80) var reachingEol = true; // invalid
 * ```
 */

var assert = require('assert');
var assign = require('lodash.assign');
var indexOf = require('lodash.indexof');

var wordlistMap = {
    english: 'wordlist-english',
    american: 'wordlist-english',
    british: 'wordlist-english',
    canadian: 'wordlist-english',
};

// Breaks names like "fooBar" into ["foo", "bar"], etc.
var reWords = (function() {
    var upper = '[A-Z\\xc0-\\xd6\\xd8-\\xde]';
    var lower = '[a-z\\xdf-\\xf6\\xf8-\\xff]+';

    return RegExp(upper + '+(?=' + upper + lower + ')|' + upper + '?' + lower + '|' + upper + '+|[0-9]+', 'g');
}());

function hasWord(dictionary, word) {
    return indexOf(dictionary, word, true) > -1;
}

function dictionariesHaveWord(dictionaries, word) {
    return dictionaries.some(function(dictionary) {
        return hasWord(dictionary, word);
    });
}

function checkWords(dictionaries, excluded, errors, words, start) {
    // Use String.prototype.replace to iterate over the words because it
    // provides an offset value useful for pinpointing a word's location.
    String(words).replace(reWords, function(word, offset) {
        word = word.toLowerCase();
        // Numbers may occasionally get orphaned; just ignore them.
        if (/^[0-9]*$/.test(word)) {
            return;
        }
        var hadWord;
        // Always ignore excluded words.
        if (hasWord(excluded, word)) {
            hadWord = false;
        } else {
            hadWord = dictionariesHaveWord(dictionaries, word);
        }
        if (!hadWord) {
            var location;
            if (offset === 0) {
                location = start;
            } else {
                location = assign({}, start);
                location.column += offset;
            }
            errors.add('Non-dictionary word "' + word + '"', location);
        }
    });
}

function isSignificantNode(node) {
    // Identifier and MemberExpression nodes always have parents, and those are
    // the only nodes that are passed to this function.
    return (
        (node.parentNode.type === 'VariableDeclarator' &&
         node.parentNode.id === node) ||
        (node.parentNode.type === 'AssignmentExpression' &&
         node.parentNode.left === node) ||
        node.parentNode.type === 'FunctionDeclaration' ||
        node.parentNode.type === 'FunctionExpression' ||
        node.parentNode.type === 'ClassDeclaration' ||
        node.parentNode.type === 'MethodDefinition' ||
        node.parentNode.type === 'LabeledStatement'
    );
}

module.exports = function() {};

module.exports.prototype = {

    configure: function(options) {
        var wasObject = typeof options === 'object' && options !== null;
        assert(
            options === true || wasObject,
            this.getOptionName() + ' option requires a true or object value ' +
                'or should be removed'
        );

        var wordDictionaries = ['english'];
        var allowedWords;
        var allowedIdentifierWords;
        var allowedPropertyWords;
        var allowedNames;
        var allowedIdentifierNames;
        var allowedPropertyNames;
        var excludedWords;

        function arrayOption(option) {
            var value = options[option];
            if (value !== undefined) {
                assert(
                    Array.isArray(value),
                    'requireDictionaryWords.' + option + ' option requires ' +
                        'an array value or should be removed'
                );
                // In order for binary searches to work the array must be
                // sorted.
                return value.sort();
            }
            return value;
        }

        if (wasObject) {
            wordDictionaries = arrayOption('dictionaries') || wordDictionaries;
            allowedWords = arrayOption('allowWords');
            allowedIdentifierWords = arrayOption('allowWordsInIdentifiers');
            allowedPropertyWords = arrayOption('allowWordsInProperties');
            allowedNames = arrayOption('allowNames');
            allowedIdentifierNames = arrayOption('allowNamesAsIdentifiers');
            allowedPropertyNames = arrayOption('allowNamesAsProperties');
            excludedWords = arrayOption('excludeWords');
        }

        this._wordDictionaries = wordDictionaries.map(function(language) {
            var packageName = wordlistMap.hasOwnProperty(language) ?
                    // Special case where one package holds many wordlists.
                    wordlistMap[language] :
                    // General case where a package is one-to-one with a
                    // wordlist.
                    'wordlist-' + language;
            var wordlist;
            try {
                wordlist = require(packageName);
            } catch (error) {
                throw new Error(
                    'Package "' + packageName + '" is not installed but required ' +
                        'for spell-checking. Go to the installation directory of ' +
                        'JSCS and run "npm install ' + packageName + '"'
                );
            }
            return wordlist[language];
        });
        if (allowedWords) {
            this._wordDictionaries.push(allowedWords);
        }

        this._identifierWordDictionaries = this._wordDictionaries.slice();
        if (allowedIdentifierWords) {
            this._identifierWordDictionaries.push(allowedIdentifierWords);
        }

        this._propertyWordDictionaries = this._wordDictionaries.slice();
        if (allowedPropertyWords) {
            this._propertyWordDictionaries.push(allowedPropertyWords);
        }

        this._nameDictionaries = [];
        if (allowedNames) {
            this._nameDictionaries.push(allowedNames);
        }

        this._identifierNameDictionaries = this._nameDictionaries.slice();
        if (allowedIdentifierNames) {
            this._identifierNameDictionaries.push(allowedIdentifierNames);
        }

        this._propertyNameDictionaries = this._nameDictionaries.slice();
        if (allowedPropertyNames) {
            this._propertyNameDictionaries.push(allowedPropertyNames);
        }

        this._excludedWords = excludedWords;
    },

    getOptionName: function() {
        return 'requireDictionaryWords';
    },

    check: function(file, errors) {
        var _this = this;

        var allowedIndex = [];

        function addToAllowedIndex(allowed, rule, namesStr, line) {
            namesStr.split(',').forEach(function(name) {
                name = name.trim();

                if (!name) {
                    return;
                }

                allowedIndex.push({
                    rule: rule,
                    name: name,
                    allowed: allowed,
                    line: line
                });
            });
        }

        function buildAllowedIndex() {
            var comments = file.getComments();
            var commentRe = new RegExp([
                '(jscs\\s*:\\s*(allow|disallow)(',
                [
                    'Words',
                    'WordsInIdentifiers',
                    'WordsInProperties',
                    'Names',
                    'NamesAsIdentifiers',
                    'NamesAsProperties'
                ].join('|'),
                '))\\s+([\\s\\S]*)',
            ].join(''));

            comments.forEach(function(comment) {
                var allowed;
                var parsed = commentRe.exec(comment.value.trim());

                if (!parsed || parsed.index !== 0) {
                    return;
                }

                allowed = parsed[2] === 'allow';
                var rule = 'allow' + parsed[3];
                var namesStr = parsed[4];
                addToAllowedIndex(allowed, rule, namesStr, comment.loc.start.line);
            }, this);
        }

        buildAllowedIndex();

        function isAllowed(context, name, line) {
            var allowed = false;
            allowedIndex.some(function(region) {
                // once the comment we're inspecting occurs after the location of the error,
                // no longer check for whether the state is allowed or disallow
                if (region.line > line) {
                    return true;
                }

                var inverseContextRe = new RegExp(
                    (context === 'identifier' ? 'Properties' : 'Identifiers') + '$'
                );

                if (inverseContextRe.test(region.rule)) {
                    return;
                }

                if (/^allowNames/.test(region.rule)) {
                    if (region.name === name) {
                        allowed = region.allowed;
                    }
                } else if (/^allowWords/.test(region.rule)) {
                    if (name.match(reWords).indexOf(region.name) > -1) {
                        allowed = region.allowed;
                    }
                }
            });

            return allowed;
        }

        function check(context, wordDictionaries, nameDictionaries, name, start) {
            if (
                isAllowed(context, name, start.line) ||
                dictionariesHaveWord(nameDictionaries, name)
            ) {
                return;
            }
            checkWords(
                wordDictionaries,
                _this._excludedWords,
                errors,
                name,
                start
            );
        }

        function checkIdentifier(name, start) {
            check(
                'identifier',
                _this._identifierWordDictionaries,
                _this._identifierNameDictionaries,
                name,
                start
            );
        }

        function checkProperty(name, start) {
            check(
                'property',
                _this._propertyWordDictionaries,
                _this._propertyNameDictionaries,
                name,
                start
            );
        }

        file.iterateNodesByType('Identifier', function(node) {
            if (!isSignificantNode(node)) {
                return;
            }
            checkIdentifier(node.name, node.loc.start);
        });

        file.iterateNodesByType('MemberExpression', function(node) {
            if (!isSignificantNode(node)) {
                return;
            }
            if (node.property.type === 'Identifier' && !node.computed) {
                checkProperty(node.property.name, node.property.loc.start);
            } else if (node.property.type === 'Literal') {
                checkProperty(node.property.value, node.property.loc.start);
            }
        });

        file.iterateNodesByType('ObjectExpression', function(node) {
            node.properties.forEach(function(prop) {
                var key = prop.key;
                if (key.type === 'Identifier') {
                    checkProperty(key.name, key.loc.start);
                } else if (key.type === 'Literal') {
                    checkProperty(key.value, key.loc.start);
                }
            });
        });
    }

};
